server/src/main/java/org/elasticsearch/transport/TcpChannel.java

```diff
@@ -118,4 +118,19 @@ static void awaitConnected(DiscoveryNode discoveryNode, List<ActionFuture<Void>>
         }
     }
 
+    /**
+     * Aborts a channel's connection with RST, then closes the channel.
+     *
+     * @param channel Channel to disconnect with RST and close
+     */
+    static void rstAndClose(TcpChannel channel) {
+        try {
+            channel.setSoLinger(0);
+        } catch (IOException | RuntimeException e) {
+            // Depending on the implementation an already closed channel can either throw an IOException or a RuntimeException
+            // that we ignore because we are closing with RST because of an error here.
```

Comment:
Which implementations throw a runtime exception?

Reply:
@jasontedor Netty4 throws a custom runtime exception here (can't catch it specifically here either because we don't have Netty on the classpath here)

Reply:
Catching and swallowing a runtime exception is pretty broad. Am I reading the code correctly that it is a channel exception? I know it is hacky, but can we check the class name manually and re-throw if it is not a io.netty.channel.ChannelException?

Reply:
@jasontedor yea, you're correct. But actually I think I just found a better fix here.

In all other implementations but the Netty one, we do this:

```
    @Override
    public void setSoLinger(int value) throws IOException {
        if (isOpen()) {
            getRawChannel().setOption(StandardSocketOptions.SO_LINGER, value);
        }
    }
```

... check if the socket is open before trying to set the setting.
Only in the Netty 4 one we don't do that and it throws in these corner cases ... I'll just add that check there I guess?

Reply:
+1